presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -729,6 +736,65 @@ public void createTable(ConnectorSession session, ConnectorTableMetadata tableMe
                 new PartitionStatistics(basicStatistics, ImmutableMap.of()));
     }
 
+    @Override
+    public ConnectorTableHandle createTemporaryTable(ConnectorSession session, List<ColumnMetadata> columns, Optional<ConnectorPartitioningMetadata> partitioning)
+    {
+        String schemaName = getTemporaryTableSchema(session);
+        String tableName = "__presto_temporary_table_" + randomUUID().toString().replaceAll("-", "_");
+        HiveStorageFormat storageFormat = getTemporaryTableStorageFormat(session);
+        Optional<HiveBucketProperty> bucketProperty = partitioning.map(partitioning1 -> {
+            HivePartitioningHandle partitioningHandle = (HivePartitioningHandle) partitioning1.getPartitioningHandle();
+            return new HiveBucketProperty(partitioning1.getPartitionColumns(), partitioningHandle.getBucketCount(), ImmutableList.of());
+        });
+        bucketProperty.ifPresent(property -> {
+            Set<String> allColumns = columns.stream()
+                    .map(ColumnMetadata::getName)
+                    .collect(toImmutableSet());
+            if (!allColumns.containsAll(property.getBucketedBy())) {
+                throw new PrestoException(INVALID_TABLE_PROPERTY, format(
+                        "Bucketing columns %s not present in schema",
+                        Sets.difference(ImmutableSet.copyOf(property.getBucketedBy()), ImmutableSet.copyOf(allColumns))));
+            }
+        });
+
+        List<HiveColumnHandle> columnHandles = getColumnHandles(columns, ImmutableSet.of(), typeTranslator);
+        storageFormat.validateColumns(columnHandles);
+
+        LocationHandle locationHandle = locationService.forNewTable(metastore, session, schemaName, tableName);
+        Path targetPath = locationService.getQueryWriteInfo(locationHandle).getTargetPath();
+
+        Table table = buildTableObject(
+                session.getQueryId(),
+                schemaName,
+                tableName,
+                session.getUser(),
+                columnHandles,
+                storageFormat,
+                ImmutableList.of(),
+                bucketProperty,
+                ImmutableMap.of(),
+                targetPath,
+                TEMPORARY_TABLE,
+                prestoVersion);
+
+        metastore.createTable(
+                session,
+                table,
+                buildInitialPrivilegeSet(table.getOwner()),
+                Optional.empty(),
+                false,
+                new PartitionStatistics(createEmptyStatistics(), ImmutableMap.of()));
+
+        try {
+            hdfsEnvironment.getFileSystem(new HdfsContext(session, table.getDatabaseName(), table.getTableName()), targetPath).mkdirs(targetPath);
```

Comment:
I must be missing something but I didn't see this `mkdir` in `createTable`. Why do we need to explicitly `mkdir` here?

Reply:
The problem is that even the permanent empty tables are not readable within transaction. As currently this is a responsibility of the metastore to create an empty directory if the table is empty. And the Metastore is called on table commit. 

Since for now we are always going to write only bucketed tables,  this problem is fixed by the previous commit, that creates bucketed files for temporary tables. I'm going to remove these lines for now. But that is something that we would probably have to fix when implementing the in transaction reads for unpartitioned tables (both, temporary and permanent).